tsql-event-store.ddl

```diff
@@ -0,0 +1,218 @@
+            )
+            BEGIN
+                THROW 50005, 'previous_id can only be null for first entity event', 1;
+            END;
+        END
+        ELSE
+        BEGIN
+            IF NOT EXISTS (
+                SELECT 1
+                FROM ledger l
+                WHERE l.event_id   = @previous_id
+                  AND l.entity     = @entity
+                  AND l.entity_key = @entity_key
+            )
+            BEGIN
+                THROW 50006, 'previous_id must be in the same entity', 1;
+            END;
+
+            IF EXISTS (
+                SELECT 1
+                FROM ledger prev
+                WHERE prev.event_id = @previous_id
+                  AND EXISTS (
+                      SELECT 1
+                      FROM ledger newer
+                      WHERE newer.entity     = @entity
+                        AND newer.entity_key = @entity_key
+                        AND newer.sequence > prev.sequence
+                  )
+            )
+            BEGIN
+                THROW 50007, 'previous_id must reference the newest event in entity', 1;
+            END;
+        END;
+
+        -- Insert
+        SET @event_id = NEWID();
+
+        INSERT INTO ledger (entity, entity_key, event, data, append_key, previous_id, event_id, timestamp)
+        VALUES (
+            @entity,
+            @entity_key,
+            @event,
+            @data,
+            @append_key,
+            @previous_id,
+            @event_id,
+            ISNULL(@timestamp, SYSDATETIMEOFFSET())
+        );
+
+        COMMIT TRAN;
+
+        -- clear session flag
+        EXEC sys.sp_set_session_context 
+            @key = N'allow_direct_ledger_insert', 
+            @value = NULL;
+    END TRY
+    BEGIN CATCH
+        IF XACT_STATE() <> 0
+            ROLLBACK TRAN;
+
+        -- always clear session flag even on error
+        EXEC sys.sp_set_session_context 
+            @key = N'allow_direct_ledger_insert', 
+            @value = NULL;
+
+        THROW;
+    END CATCH;
+END;
+GO
+
+
+CREATE OR ALTER TRIGGER no_direct_insert_ledger
+ON ledger
+AFTER INSERT
+AS
+BEGIN
+    SET NOCOUNT ON;
+    SET XACT_ABORT ON;
+
+    IF SESSION_CONTEXT(N'allow_direct_ledger_insert') = 1
+        RETURN;
+
+    THROW 50003, 'Use append_event procedure to insert events into the ledger', 1;
+END;
+GO
+
+CREATE OR ALTER VIEW replay_events AS
+SELECT
+    entity,
+    entity_key,
+    event,
+    data,
+    append_key,
+    previous_id,
+    event_id,
+    timestamp,
+    sequence
+FROM ledger;
+GO
+
+
+CREATE OR ALTER FUNCTION replay_events_after(@after_event_id UNIQUEIDENTIFIER)
+RETURNS TABLE
+AS
+RETURN
+(
+    SELECT 
+        entity,
+        entity_key,
+        event,
+        data,
+        append_key,
+        previous_id,
+        event_id,
+        timestamp,
+        sequence
+    FROM ledger
+    WHERE sequence > (
+        SELECT sequence 
+        FROM ledger 
+        WHERE event_id = @after_event_id
+    )
```

Comment:
Does this need `ORDER BY sequence`? Practically, it probably doesn't given how events are written to the database, but I put it in the other versions to guarantee ordering.